File: sdks/python/src/boxsdk/models/shared/folder.py
import dataclasses
from datetime import date, datetime
from marshmallow import fields
import dateutil.parser
from typing import Any,Optional
from enum import Enum
from dataclasses_json import dataclass_json
from boxsdk import utils
from ..shared import file_mini as shared_file_mini
from ..shared import folder_mini as shared_folder_mini
from ..shared import weblink_mini as shared_weblink_mini

class FolderUserBaseTypeEnum(str, Enum):
    USER = "user"


@dataclass_json
@dataclasses.dataclass
class FolderUserBase:
    r"""FolderUserBase
    The user who created this folder
    """
    
    login: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('login') }})
    name: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('name') }})
    type: FolderUserBaseTypeEnum = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('type') }})
    id: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('id') }})
    
class FolderFolderUploadEmailAccessEnum(str, Enum):
    OPEN = "open"
    COLLABORATORS = "collaborators"


@dataclass_json
@dataclasses.dataclass
class FolderFolderUploadEmail:
    access: Optional[FolderFolderUploadEmailAccessEnum] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('access') }})
    email: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('email') }})
    
class FolderItemsOrderDirectionEnum(str, Enum):
    ASC = "ASC"
    DESC = "DESC"


@dataclass_json
@dataclasses.dataclass
class FolderItemsOrder:
    r"""FolderItemsOrder
    The order in which a pagination is ordered
    """
    
    by: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('by') }})
    direction: Optional[FolderItemsOrderDirectionEnum] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('direction') }})
    

@dataclass_json
@dataclasses.dataclass
class FolderItems:
    r"""FolderItems
    A page of the items that are in the folder.
    
    This field can only be requested when querying a folder's
    information, not when querying a folder's items.
    """
    
    entries: Optional[list[Any]] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('entries') }})
    limit: Optional[int] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('limit') }})
    offset: Optional[int] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('offset') }})
    order: Optional[list[FolderItemsOrder]] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('order') }})
    total_count: Optional[int] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('total_count') }})
    
class FolderItemStatusEnum(str, Enum):
    ACTIVE = "active"
    TRASHED = "trashed"
    DELETED = "deleted"

class FolderFolderBaseTypeEnum(str, Enum):
    FOLDER = "folder"


@dataclass_json
@dataclasses.dataclass
class FolderFolderBase:
    r"""FolderFolderBase
    The optional folder that this folder is located within.
    
    This value may be `null` for some folders such as the
    root folder or the trash folder.
    """
    
    id: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('id') }})
    type: FolderFolderBaseTypeEnum = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('type') }})
    etag: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('etag') }})
    name: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('name') }})
    sequence_id: Optional[dict[str, Any]] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('sequence_id') }})
    

@dataclass_json
@dataclasses.dataclass
class FolderPathCollection:
    r"""FolderPathCollection
    The tree of folders that this folder is contained in,
    starting at the root.
    """
    
    entries: list[shared_folder_mini.FolderMini] = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('entries') }})
    total_count: int = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('total_count') }})
    
class FolderSharedLinkAccessEnum(str, Enum):
    OPEN = "open"
    COMPANY = "company"
    COLLABORATORS = "collaborators"

class FolderSharedLinkEffectiveAccessEnum(str, Enum):
    OPEN = "open"
    COMPANY = "company"
    COLLABORATORS = "collaborators"

class FolderSharedLinkEffectivePermissionEnum(str, Enum):
    CAN_EDIT = "can_edit"
    CAN_DOWNLOAD = "can_download"
    CAN_PREVIEW = "can_preview"
    NO_ACCESS = "no_access"


@dataclass_json
@dataclasses.dataclass
class FolderSharedLinkPermissions:
    r"""FolderSharedLinkPermissions
    Defines if this link allows a user to preview, edit, and download an item.
    These permissions refer to the shared link only and
    do not supersede permissions applied to the item itself.
    """
    
    can_download: bool = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('can_download') }})
    can_edit: bool = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('can_edit') }})
    can_preview: bool = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('can_preview') }})
    

@dataclass_json
@dataclasses.dataclass
class FolderSharedLink:
    r"""FolderSharedLink
    The shared link for this folder. This will be
    `null` if no shared link has been created for this
    folder.
    """
    
    download_count: int = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('download_count') }})
    effective_access: FolderSharedLinkEffectiveAccessEnum = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('effective_access') }})
    effective_permission: FolderSharedLinkEffectivePermissionEnum = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('effective_permission') }})
    is_password_enabled: bool = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('is_password_enabled') }})
    preview_count: int = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('preview_count') }})
    url: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('url') }})
    access: Optional[FolderSharedLinkAccessEnum] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('access') }})
    download_url: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('download_url') }})
    permissions: Optional[FolderSharedLinkPermissions] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('permissions') }})
    unshared_at: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('unshared_at'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    vanity_name: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('vanity_name') }})
    vanity_url: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('vanity_url') }})
    
class FolderTypeEnum(str, Enum):
    FOLDER = "folder"


@dataclass_json
@dataclasses.dataclass
class Folder:
    r"""Folder
    The bare basic representation of a folder, the minimal
    amount of fields returned when using the `fields` query
    parameter.
    """
    
    created_by: FolderUserBase = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('created_by') }})
    id: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('id') }})
    item_collection: FolderItems = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('item_collection') }})
    item_status: FolderItemStatusEnum = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('item_status') }})
    modified_by: FolderUserBase = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('modified_by') }})
    owned_by: FolderUserBase = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('owned_by') }})
    path_collection: FolderPathCollection = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('path_collection') }})
    size: int = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('size') }})
    type: FolderTypeEnum = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('type') }})
    content_created_at: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('content_created_at'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    content_modified_at: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('content_modified_at'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    created_at: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('created_at'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    description: Optional[dict[str, Any]] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('description') }})
    etag: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('etag') }})
    folder_upload_email: Optional[FolderFolderUploadEmail] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('folder_upload_email') }})
    modified_at: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('modified_at'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    name: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('name') }})
    parent: Optional[FolderFolderBase] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('parent') }})
    purged_at: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('purged_at'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    sequence_id: Optional[dict[str, Any]] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('sequence_id') }})
    shared_link: Optional[FolderSharedLink] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('shared_link') }})
    trashed_at: Optional[datetime] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('trashed_at'), 'encoder': utils.datetimeisoformat(True), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    

File: ts-sdk/src/sdk/models/shared/folder.ts
import { SpeakeasyMetadata, SpeakeasyBase } from "../../../internal/utils";
import { FileMini } from "./filemini";
import { FolderMini } from "./foldermini";
import { WebLinkMini } from "./weblinkmini";


export enum FolderUserBaseTypeEnum {
    User = "user"
}


// FolderUserBase
/** 
 * The user who created this folder
**/
export class FolderUserBase extends SpeakeasyBase {
  @SpeakeasyMetadata({ data: "json, name=id" })
  id?: string;

  @SpeakeasyMetadata({ data: "json, name=login" })
  login: string;

  @SpeakeasyMetadata({ data: "json, name=name" })
  name: string;

  @SpeakeasyMetadata({ data: "json, name=type" })
  type: FolderUserBaseTypeEnum;
}

export enum FolderFolderUploadEmailAccessEnum {
    Open = "open",
    Collaborators = "collaborators"
}


export class FolderFolderUploadEmail extends SpeakeasyBase {
  @SpeakeasyMetadata({ data: "json, name=access" })
  access?: FolderFolderUploadEmailAccessEnum;

  @SpeakeasyMetadata({ data: "json, name=email" })
  email?: string;
}

export enum FolderItemsOrderDirectionEnum {
    Asc = "ASC",
    Desc = "DESC"
}


// FolderItemsOrder
/** 
 * The order in which a pagination is ordered
**/
export class FolderItemsOrder extends SpeakeasyBase {
  @SpeakeasyMetadata({ data: "json, name=by" })
  by?: string;

  @SpeakeasyMetadata({ data: "json, name=direction" })
  direction?: FolderItemsOrderDirectionEnum;
}


// FolderItems
/** 
 * A page of the items that are in the folder.
 * 
 * This field can only be requested when querying a folder's
 * information, not when querying a folder's items.
**/
export class FolderItems extends SpeakeasyBase {
  @SpeakeasyMetadata({ data: "json, name=entries" })
  entries?: any[];

  @SpeakeasyMetadata({ data: "json, name=limit" })
  limit?: number;

  @SpeakeasyMetadata({ data: "json, name=offset" })
  offset?: number;

  @SpeakeasyMetadata({ data: "json, name=order", elemType: FolderItemsOrder })
  order?: FolderItemsOrder[];

  @SpeakeasyMetadata({ data: "json, name=total_count" })
  totalCount?: number;
}

export enum FolderItemStatusEnum {
    Active = "active",
    Trashed = "trashed",
    Deleted = "deleted"
}

export enum FolderFolderBaseTypeEnum {
    Folder = "folder"
}


// FolderFolderBase
/** 
 * The optional folder that this folder is located within.
 * 
 * This value may be `null` for some folders such as the
 * root folder or the trash folder.
**/
export class FolderFolderBase extends SpeakeasyBase {
  @SpeakeasyMetadata({ data: "json, name=etag" })
  etag?: string;

  @SpeakeasyMetadata({ data: "json, name=id" })
  id: string;

  @SpeakeasyMetadata({ data: "json, name=name" })
  name?: string;

  @SpeakeasyMetadata({ data: "json, name=sequence_id" })
  sequenceId?: Record<string, any>;

  @SpeakeasyMetadata({ data: "json, name=type" })
  type: FolderFolderBaseTypeEnum;
}


// FolderPathCollection
/** 
 * The tree of folders that this folder is contained in,
 * starting at the root.
**/
export class FolderPathCollection extends SpeakeasyBase {
  @SpeakeasyMetadata({ data: "json, name=entries", elemType: FolderMini })
  entries: FolderMini[];

  @SpeakeasyMetadata({ data: "json, name=total_count" })
  totalCount: number;
}

export enum FolderSharedLinkAccessEnum {
    Open = "open",
    Company = "company",
    Collaborators = "collaborators"
}

export enum FolderSharedLinkEffectiveAccessEnum {
    Open = "open",
    Company = "company",
    Collaborators = "collaborators"
}

export enum FolderSharedLinkEffectivePermissionEnum {
    CanEdit = "can_edit",
    CanDownload = "can_download",
    CanPreview = "can_preview",
    NoAccess = "no_access"
}


// FolderSharedLinkPermissions
/** 
 * Defines if this link allows a user to preview, edit, and download an item.
 * These permissions refer to the shared link only and
 * do not supersede permissions applied to the item itself.
**/
export class FolderSharedLinkPermissions extends SpeakeasyBase {
  @SpeakeasyMetadata({ data: "json, name=can_download" })
  canDownload: boolean;

  @SpeakeasyMetadata({ data: "json, name=can_edit" })
  canEdit: boolean;

  @SpeakeasyMetadata({ data: "json, name=can_preview" })
  canPreview: boolean;
}


// FolderSharedLink
/** 
 * The shared link for this folder. This will be
 * `null` if no shared link has been created for this
 * folder.
**/
export class FolderSharedLink extends SpeakeasyBase {
  @SpeakeasyMetadata({ data: "json, name=access" })
  access?: FolderSharedLinkAccessEnum;

  @SpeakeasyMetadata({ data: "json, name=download_count" })
  downloadCount: number;

  @SpeakeasyMetadata({ data: "json, name=download_url" })
  downloadUrl?: string;

  @SpeakeasyMetadata({ data: "json, name=effective_access" })
  effectiveAccess: FolderSharedLinkEffectiveAccessEnum;

  @SpeakeasyMetadata({ data: "json, name=effective_permission" })
  effectivePermission: FolderSharedLinkEffectivePermissionEnum;

  @SpeakeasyMetadata({ data: "json, name=is_password_enabled" })
  isPasswordEnabled: boolean;

  @SpeakeasyMetadata({ data: "json, name=permissions" })
  permissions?: FolderSharedLinkPermissions;

  @SpeakeasyMetadata({ data: "json, name=preview_count" })
  previewCount: number;

  @SpeakeasyMetadata({ data: "json, name=unshared_at" })
  unsharedAt?: Date;

  @SpeakeasyMetadata({ data: "json, name=url" })
  url: string;

  @SpeakeasyMetadata({ data: "json, name=vanity_name" })
  vanityName?: string;

  @SpeakeasyMetadata({ data: "json, name=vanity_url" })
  vanityUrl?: string;
}

export enum FolderTypeEnum {
    Folder = "folder"
}


// Folder
/** 
 * The bare basic representation of a folder, the minimal
 * amount of fields returned when using the `fields` query
 * parameter.
**/
export class Folder extends SpeakeasyBase {
  @SpeakeasyMetadata({ data: "json, name=content_created_at" })
  contentCreatedAt?: Date;

  @SpeakeasyMetadata({ data: "json, name=content_modified_at" })
  contentModifiedAt?: Date;

  @SpeakeasyMetadata({ data: "json, name=created_at" })
  createdAt?: Date;

  @SpeakeasyMetadata({ data: "json, name=created_by" })
  createdBy: FolderUserBase;

  @SpeakeasyMetadata({ data: "json, name=description" })
  description?: Record<string, any>;

  @SpeakeasyMetadata({ data: "json, name=etag" })
  etag?: string;

  @SpeakeasyMetadata({ data: "json, name=folder_upload_email" })
  folderUploadEmail?: FolderFolderUploadEmail;

  @SpeakeasyMetadata({ data: "json, name=id" })
  id: string;

  @SpeakeasyMetadata({ data: "json, name=item_collection" })
  itemCollection: FolderItems;

  @SpeakeasyMetadata({ data: "json, name=item_status" })
  itemStatus: FolderItemStatusEnum;

  @SpeakeasyMetadata({ data: "json, name=modified_at" })
  modifiedAt?: Date;

  @SpeakeasyMetadata({ data: "json, name=modified_by" })
  modifiedBy: FolderUserBase;

  @SpeakeasyMetadata({ data: "json, name=name" })
  name?: string;

  @SpeakeasyMetadata({ data: "json, name=owned_by" })
  ownedBy: FolderUserBase;

  @SpeakeasyMetadata({ data: "json, name=parent" })
  parent?: FolderFolderBase;

  @SpeakeasyMetadata({ data: "json, name=path_collection" })
  pathCollection: FolderPathCollection;

  @SpeakeasyMetadata({ data: "json, name=purged_at" })
  purgedAt?: Date;

  @SpeakeasyMetadata({ data: "json, name=sequence_id" })
  sequenceId?: Record<string, any>;

  @SpeakeasyMetadata({ data: "json, name=shared_link" })
  sharedLink?: FolderSharedLink;

  @SpeakeasyMetadata({ data: "json, name=size" })
  size: number;

  @SpeakeasyMetadata({ data: "json, name=trashed_at" })
  trashedAt?: Date;

  @SpeakeasyMetadata({ data: "json, name=type" })
  type: FolderTypeEnum;
}
